Allow domain to add their own banners

mkangia opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Let projects add their own banners which will be shown to their users on HQ each time they log-in.
This is based on same implementation as the maintenance alerts set on HQ i.e projects can enable/disable an alert and user can dismiss them once they see it and the alerts will be show again 7 days after the user dismisses the last visible alert.
With one difference being that alerts will be visible again when user logs in.
Additionally, a delete option has also been added to remove an alert as well.

This has been put behind a feature flag for now but since this has been asked for, by projects in other divisions, this could be made GA soon. More refinements can be expected for better UX when that is done.

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Review by commit :fish: 
https://dimagi-dev.atlassian.net/browse/SC-3003

Building on the model and implementation available for maintenance alerts, this PR extends that model to support alert created by domains.
For audit purposes, it is also tracked who added the alert, though that information is lost if alert is removed.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
CUSTOM_DOMAIN_BANNER_ALERTS

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Apart from behind a feature flag,
Tested locally or taken care that
1. domain alerts are not visible on other domains [[Code Ref](https://github.com/dimagi/commcare-hq/pull/33592/commits/8371b87d030f055b7e77995ab3f2f516d8d54a6a#diff-d0f5be95e8abb64eb7341b9017bed86f57b8c24bbff6d41985b5894dc3188ad3R546)]
2. one domain can't add alerts for another domain since it is set in code and not by user [[Code Ref](https://github.com/dimagi/commcare-hq/pull/33592/commits/8371b87d030f055b7e77995ab3f2f516d8d54a6a#diff-d0f5be95e8abb64eb7341b9017bed86f57b8c24bbff6d41985b5894dc3188ad3R546)]
3. superusers can't view/modify alerts added by domain [[[Code Ref](https://github.com/dimagi/commcare-hq/pull/33592/commits/76bad649dfa46396e452d57fc939b6f5929f699f)]
4. active alerts are cached for one minute, so a deleted/deactivated alert would stay for a minute, which should be acceptable

HQ already ensures that [only relevant alerts are shown on a page on HQ](https://github.com/dimagi/commcare-hq/blob/8e5d820b9a3002b9a6354c68a754575c4503f7af/corehq/apps/hqwebapp/templatetags/hq_shared_tags.py#L397-L398), so that takes care of any edge cases.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
QA will be requested

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [X] The migrations in this code can be safely applied first independently of the code
~~https://github.com/dimagi/commcare-hq/pull/33595 makes this possible~~

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

~~This PR renames a table. So in case it is reverted, the table should be considered to be renamed back. 
It is not a blocker if the table is not immediately renamed. In case there are no maintenance alerts on the website, there is no need to rename the table back immediately and can be delayed till any maintenance alert needs to be added or till the PR can be added back.~~
Only thing to consider is that if there are any alerts that have been added by any domain.
`Alert.objects.filter(created_by_domain__isnull=False)`
Such alerts should be removed (preferred) or deactivated by setting there `active` to be `False` to buy more time. The project should be notified that their alert has been removed.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

mkangia committed:
adds view & form for new alert

mkangia committed:
renames model to a better name

mkangia committed:
saves alert

mkangia committed:
lists alerts

mkangia committed:
adds ability to set alert status

mkangia committed:
tracks closed domain alerts in cookie

mkangia committed:
resets cookie on login

mkangia committed:
extracts setUpAlert function

mkangia committed:
restricts admin view to global alerts

mkangia committed:
adds feature flag

mkangia committed:
checks for toggle on view and link to view

mkangia committed:
tracks user for audit

mkangia committed:
merges migrations

mkangia committed:
adds model to be deleted with domain deletion

mkangia committed:
updates repr function with new name

mkangia committed:
renames function

mkangia committed:
updates bootstrap diffs

mkangia committed:
lint

mkangia committed:
adds alert deletion

mkangia committed:
adds list view as default response

mkangia committed:
refactors post method

mkangia commented on corehq/apps/domain/forms.py:
```diff
@@ -2655,3 +2655,28 @@ def clean_version(self):
+        label="Text",
+        widget=forms.Textarea,
+        required=True,
+    )
```
Keep it simple and support only text for now. 
This can be extended to add a start/end time as well which is already available on the model and is used for maintenance alerts

mkangia commented on corehq/tabs/tabclasses.py:
```diff
@@ -1996,10 +1997,16 @@ def _get_administration_section(domain):
+        },
+        {
+            'title': _(ManageDomainAlertsView.page_title),
+            'url': reverse(ManageDomainAlertsView.urlname, args=[domain])
```
moved behind a feature flag later.

mkangia commented on corehq/apps/domain/views/settings.py:
```diff
@@ -540,7 +541,11 @@ def form(self):
-            # ToDo: save the alert
+            CommCareHQAlert.objects.create(
+                created_by_domain=self.domain,
+                domains=[self.domain],
```
This is necessary despite setting created_by_domain to let HQ know the relevant domain for the alert during filtering for UI.

mkangia commented on corehq/apps/domain/templates/domain/admin/manage_alerts.html:
```diff
@@ -1,8 +1,50 @@
     <form method="post">
       {% crispy form %}
     </form>
+    <div id="ko-alert-container">
```
Template kept mostly same as that currently available for superusers

mkangia commented on corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/main.js:
```diff
@@ -409,15 +409,28 @@ hqDefine('hqwebapp/js/bootstrap3/main', [
             var alertCookie = 'alerts_maintenance';
             var closedAlerts = $.cookie(alertCookie) ? JSON.parse($.cookie(alertCookie)) : [];
 
+            var viewedDomainAlertsCookie = 'viewed_domain_alerts';
```
Used a separate cookie for domain alerts instead of extending the one available just above, since this needs to be reset on login.

mkangia commented on corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/main.js:
```diff
@@ -409,15 +409,28 @@ hqDefine('hqwebapp/js/bootstrap3/main', [
+                                closedAlerts.push(id);
+                                $.cookie(alertCookie, JSON.stringify(closedAlerts), { expires: 7, path: '/', secure: initialPageData.get('secure_cookies') });
+                            });
+                        }
```
later refactored by extracting a function to avoid duplication.

mkangia commented on migrations.lock:
```diff
@@ -588,8 +588,7 @@ hqwebapp
  0011_rename_maintenancealert_commcarehqalert
- 0012_commcarehqalert_created_by_domain
- 0013_commcarehqalert_created_by_user
+ 0012_auto_20231006_1052
```
Didn't merge further with `0011_rename_maintenancealert_commcarehqalert` considering that migration was doing a rename which felt like it deserved to be kept as a single migration.

mkangia replied:
renaming was avoided due to avoid unexpected issues during deploy. Only new fields were added and the "model" name was changed

mkangia commented on corehq/apps/hqwebapp/tasks.py:
```diff
@@ -242,11 +242,13 @@ def clean_expired_transient_emails():
         )
 
 
+# ToDo: remove/modify this if the related metric is removed/retained
```
Got a confirmation from SaaS (Danny) that this was okay to remove. So this was removed later.

mkangia commented:
Hold off on merge till https://github.com/dimagi/commcare-hq/pull/33595 is deployed.

mkangia committed:
retains old db table & recreates migrations

mkangia committed:
refactors to avoid branch conflicts for staging

kaapstorm commented on corehq/apps/hqwebapp/models.py:
```diff
@@ -35,6 +35,7 @@ class CommCareHQAlert(models.Model):
 
     text = models.TextField()
     domains = ArrayField(models.CharField(max_length=126), null=True)
+    created_by_domain = models.CharField(max_length=255, null=True, db_index=True)
```
If we already have a nullable `domains` field, do we need a `created_by_domain` field?

We can filter global alerts using `domains__isnull=True`. In templates we can use `{{ alerts.domains|BOOL }}`.

Could we just allow an admin to manage alerts if they are a member of all domains in `domains`?

mkangia replied:
Hello,

This is to identify if an alert was added "by" the domain and then let domain alter it.
You are right about filtering for global alerts, though we can't use the existing field to filter for alerts created by domain.
It is possible to superuser to add alerts for a domain as well, which populates the "domains" column with that domain.

> Could we just allow an admin to manage alerts if they are a member of all domains in domains?

Didn't think about that. Though if needed in future, current implementation can be extended for this.

kaapstorm commented on corehq/apps/domain/views/settings.py:
```diff
@@ -559,3 +569,19 @@ def post(self, request, *args, **kwargs):
+            alert = CommCareHQAlert.objects.get(
+                created_by_domain=self.domain,
+                id=alert_id
+            )
```
If you were hypothetically to use `domains` instead of `created_by_domain`, you could fetch the instance like this:
```suggestion
            alert = CommCareHQAlert.objects.filter(
                domains__contains=[self.domain]
            ).get(id=alert_id)
```

mkangia replied:
Replied on the other [comment](https://github.com/dimagi/commcare-hq/pull/33592#discussion_r1352322257) that created_by_domain is needed to correctly identify alerts created by the domain.

kaapstorm reviewed (approved):
This looks great! My question about whether we can just use the existing `domains` field without adding `created_by_domain` is not blocking.

mkangia committed:
adds a comment to function

This is done to change function definition to avoid using stale cache with model objects with old class name

mkangia commented:
> Hold off on merge till https://github.com/dimagi/commcare-hq/pull/33595 is deployed.

This got a bit hazy in terms of it being successful so decided to retain the original table name itself.
So now we should be unblocked for this PR. Will update the description.

mkangia commented on corehq/apps/hqwebapp/migrations/0011_auto_20231010_0950.py:
```diff
@@ -0,0 +1,31 @@
+            old_name='MaintenanceAlert',
+            new_name='CommCareHQAlert',
+        ),
+    ]
```
Just to share what this would run
```
$ ./manage.py sqlmigrate hqwebapp 0011
corehq.util.log.HqAdminEmailHandler logger is being changed to logging.StreamHandler
BEGIN;
--
-- Add field created_by_domain to maintenancealert
--
ALTER TABLE "hqwebapp_maintenancealert" ADD COLUMN "created_by_domain" varchar(255) NULL;
--
-- Add field created_by_user to maintenancealert
--
ALTER TABLE "hqwebapp_maintenancealert" ADD COLUMN "created_by_user" varchar(128) NULL;
--
-- Rename table for maintenancealert to hqwebapp_maintenancealert
--
--
-- Rename model MaintenanceAlert to CommCareHQAlert
--
CREATE INDEX "hqwebapp_maintenancealert_created_by_domain_c68ad7b8" ON "hqwebapp_maintenancealert" ("created_by_domain");
CREATE INDEX "hqwebapp_maintenancealert_created_by_domain_c68ad7b8_like" ON "hqwebapp_maintenancealert" ("created_by_domain" varchar_pattern_ops);
COMMIT;
```

millerdev commented on corehq/apps/domain/views/settings.py:
```diff
@@ -517,3 +520,87 @@ class ManageDomainMobileWorkersView(ManageMobileWorkersMixin, BaseAdminProjectSe
+            self._create_alert()
+            messages.success(request, _("Alert saved!"))
+        else:
+            messages.error(request, _("There was an error saving your alert. Please try again!"))
```
Will additional details about the error(s) be displayed in addition to this message? Simply trying again without changing the submitted values in some way is likely to produce the same error.

The same applies below for unsuccessful `_update_alert`

mkangia replied:
this last else block is just in case someone does something funky on the page or for rare unexpected scenarios.
If the forms are used as intended, there will either be a "command" (activate/deactivate) or the form with the text field (which is a required field on the form). Errors are really not expected on this page as for how it is now.
Hence the generic message. Open to suggestions if I am being too positive.

millerdev replied:
The typical purpose of form validation is to construct a detailed list of errors to be displayed to the user so they can be corrected. This form does not seem to be following that pattern. Since this would only happen if "someone does something funky on the page or for rare unexpected scenarios", I suggest simply raising an exception so we can find the issue in Sentry and have more information for debugging.
```suggestion
            raise ValidationError(form.errors.as_text())
```

mkangia replied:
> The typical purpose of form validation is to construct a detailed list of errors to be displayed to the user so they can be corrected

Right. Not sure if I misguided earlier. The error messages related to the form will be shown within the form itself. This error message is visible on the top as a generic message. Here is how it will look like.

![Screenshot from 2023-10-11 14-42-03](https://github.com/dimagi/commcare-hq/assets/3864163/4ec2dbc2-965d-4c76-a97c-a4eae539c3c1)

millerdev replied:
Great. My original question here was if specific validation errors are displayed for each field:

> Will additional details about the error(s) be displayed in addition to this message?

I now see the answer is "Yes." Thanks for the screenshot. No change is required here.

mkangia replied:
Cool, sorry about the confusion.

millerdev commented on corehq/apps/domain/views/settings.py:
```diff
@@ -517,3 +520,87 @@ class ManageDomainMobileWorkersView(ManageMobileWorkersMixin, BaseAdminProjectSe
+
+    def _apply_command(self, request, command):
+        alert_id = request.POST.get('alert_id')
+        if alert_id:
```
Should there be an error if this is false? As is it will silently do nothing, which might be unexpected.

mkangia replied:
Alert ID is passed by JS. It should not be missing.
This is just to avoid errors in case someone messes with the request.

millerdev replied:
I think it's better to fail loudly. From the Zen

```
$ python -m this | grep silently
Errors should never pass silently.
```

mkangia replied:
Not sure I would count this as an error.
I would think here is that, someone wanted to break HQ, so attempts to send the same request but without an alert ID. So we simply not do anything.
If it was possible for user to do this from UI somehow, then I'd consider showing a message that alert ID was missing.

millerdev replied:
Since there is no way to trigger this action in the UI, and it's never expected for a request to be submitted without `alert_id`, then it's better to not add code to handle that scenario.
```suggestion
        assert alert_id
```
It's good to fail early, but please don't silently ignore an unexpected state in code. Silent failures like that make it much more difficult to figure out why things are not working if that condition is triggered for some unforeseen reason.

mkangia replied:
> Silent failures like that make it much more difficult to figure out why things are not working if that condition is triggered for some unforeseen reason.

Good point. Noted.

(Will make the change in code instead of using the commit suggestion here to account for other changes that are needed along)

So, take away here would be to at least fail hard for rare/unexpected scenarios. One could even let the user know though its not worth adding that logic branch in the code. So, for example, here, we had two options
1. raise an exception if alert_id is missing
2. return with an error message "Alert ID not passed"
In terms of UX 2 is better than 1, but code wise that adds an unnecessary "if" condition, so we go with 1 or we should go with 2?

Also, I would think of exceptions as for "unhandled" scenarios. So if an exception is expected, it should be accounted (rescued) for and handled gracefully. Hence, I am in split about using an assertion here.

millerdev replied:
> take away here would be to at least fail hard for rare/unexpected scenarios

Yes, exactly. In this case you have said

> It should not be missing. This is just to avoid errors in case someone messes with the request.

To me that would indicate it was a rare/unexpected scenario, and it is probably not worth adding special error handling for that, so go with option 1. If you later notice, while monitoring Sentry after rolling out the change, that this unexpected condition is triggered then it would be worth it to figure out why it is happening and fix the issue at that time.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/33592/commits/7feb32bb00bd9d78db8613cfa658e8650aab8e6e

millerdev commented on corehq/apps/domain/views/settings.py:
```diff
@@ -517,3 +520,87 @@ class ManageDomainMobileWorkersView(ManageMobileWorkersMixin, BaseAdminProjectSe
+            if command == 'activate':
+                alert.active = True
+            elif command == 'deactivate':
+                alert.active = False
```
What if `command` is something other than `activate` or `deactivate`? Should that trigger an error?

mkangia replied:
Same as https://github.com/dimagi/commcare-hq/pull/33592/files#r1352812929

Some of these checks are to avoid errors/exceptions on HQ in case someone attempts to do something unexpected and not worry about showing them a proper error message but just ignoring that request with least possible effort.

millerdev replied:
> these checks are to avoid errors/exceptions on HQ in case someone attempts to do something unexpected

As stated above, code should not allow unexpected conditions to pass silently. We _WANT_ errors and exceptions if someone tries to do something unexpected so we can debug that situation.

mkangia replied:
Added check for expected commands and to share an error in case unexpected command is received
https://github.com/dimagi/commcare-hq/pull/33592/commits/085ee5f0125f7fcb470044e2086cdd6e8f60f2ec

millerdev commented on corehq/apps/domain/views/settings.py:
```diff
@@ -517,3 +520,87 @@ class ManageDomainMobileWorkersView(ManageMobileWorkersMixin, BaseAdminProjectSe
+                id=alert_id
+            )
+        except CommCareHQAlert.DoesNotExist:
+            return None
```
The functions that call this will simply do nothing if this returns `None`, which seems like an unexpected outcome from the user's perspective. What do you think of communicating this error the the user?

mkangia replied:
One genuine possible edge case is, if a user loads the page and tries to update an alert that is deleted by another user in the interim.
I can update the error message to say "Alert not found" if that seems like worth it.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/33592/commits/73f9a875ba16b83557c65ba5e406843a626dd86d

millerdev commented on corehq/apps/hqwebapp/models.py:
```diff
@@ -24,7 +24,7 @@ def __new__(cls, category, action, label=None):
 
 
-class MaintenanceAlert(models.Model):
+class CommCareHQAlert(models.Model):
```
Seems like most classes in HQ could have a `CommCareHQ` prefix, and that would be redundant. What do you think of renaming this to `Alert`?

Aside: IMO the model rename doesn't seem particularly important, and it introduced a lot of noise in the change set.

mkangia replied:
Agreed I could lose the 'CommCareHQ" in the name.
I think this could be better called "Notice" or "MessageBanner". I feel alert has a sense of "alarm" that these notices would not always be. Any suggestions?
We already have notification model.

> the model rename doesn't seem particularly important,

Okay. I do think the rename was important/relevant since for domains it has nothing to do with maintenance. I didn't consider the issues it would create and went for it. Renaming a model is definitely not straight forward.

mkangia replied:
> Did you decide not to rename this?

Not really. Was waiting for some follow up feedback on this.

> I think this could be better called "Notice" or "MessageBanner". I feel alert has a sense of "alarm" that these notices would not always be. Any suggestions?

I am unsure about using "Alert' because of the reason above. I think this feature is closer to a "banner" than an "alert", for instance, the project requesting this feature wants to show the user a data privacy message to the user each time they login. If that does not seem logical, I am happy to rename it to `Alert` only.

> Seems like most classes in HQ could have a CommCareHQ prefix, and that would be redundant.

Right. It does look like we have though, built a series of such things. There are numerous models with `CommCare` prefix.

millerdev replied:
> Was waiting for some follow up feedback on this.

I don't have strong opinions on the name as long as it does not have the redundant prefix. `Alert` seems fine to me since it does create a pop up that is intended to alert the user to something. It seems better for the name to imply a bit of alarm rather than to downplay the fact that this might be annoying to the user (who likes popups).

> It does look like we have though, built a series of such things. There are numerous models with `CommCare` prefix.

I do not think this is a good pattern to follow. It violates the DRY principle.

mkangia replied:
Cool,

Renamed it to Alert
https://github.com/dimagi/commcare-hq/pull/33592/commits/c2856a76caa79035d474df6fdfd54161fb704e31

One follow up,
I am thinking I should merge this new migration with the old one by deleting the new one and just modify the old one.
Though the old migration has been run on staging, so my plan is to let the new migration run on staging once and then do the merge/remove business.
I can't rollback the migration on staging because there are already some domain alerts added by QA for testing which would get messed up.
Just checking if that sounds like the right approach.

millerdev replied:
Yes, please squash all migrations in this PR into a single migration.

The plan for staging sounds reasonable as long as the resulting database schema does not change as a result of squashing, which it should not. You may need to fake the new squashed migration on staging since it will have already been deployed. Also note that the application of the new squashed migration will not be tested on staging if you do it this way.

millerdev commented on corehq/apps/hqwebapp/models.py:
```diff
@@ -35,25 +35,29 @@ class MaintenanceAlert(models.Model):
 
     def __repr__(self):
-        return "MaintenanceAlert(text='{}', active='{}', domains='{}')".format(
+        return "CommCareHQAlert(text='{}', active='{}', domains='{}')".format(
             self.text, self.active, ", ".join(self.domains) if self.domains else "All Domains")
```
nit
```suggestion
        domains = ", ".join(self.domains) if self.domains else "All Domains"
        return f"{type(self).__name__}(text={self.text!r}, active={self.active!r}, domains={domains})"
```
Note intentional omission of `!r` on `domains` (and also no enclosing quotes) since the value of that attribute does not match what is printed in this representation, and what is printed is not a proper Python data structure that could be passed in to reconstruct this object.

mkangia replied:
So instead of 
`CommCareHQAlert(text='Maintenance alert', active='False', domains='All Domains'`
this would now be
`CommCareHQAlert(text='Maintenance alert', active=False, domains=All Domains)`
or
`CommCareHQAlert(text='Maintenance alert', active=False, domains=test, test2)`

Confirming that this is what you expected how this would look like?

> what is printed is not a proper Python data structure that could be passed in to reconstruct this object.

Yup. For True/False this change seems okay, though seeing a comma separated list of domains without the enclosing quotes does look weird. And since we have used quotes for the text, it makes sense to use it for other string representations as well.
Or if we don't join domains, it would be
`CommCareHQAlert(text='Maintenance alert', active=False, domains=['test', 'test2'])`
which makes a bit more sense to me.

millerdev replied:
> Confirming that this is what you expected how this would look like?

Yes.

> seeing a comma separated list of domains without the enclosing quotes does look weird

Maybe a little, but it communicates the value and it's obvious that that value cannot be used to reconstruct the object, unlike a properly formatted string, which would be misleading because it is not a valid value that can be used to reconstruct the object.

> if we don't join domains, it would be... `domains=['test', 'test2']`

That would be fine if you like that better. The down-side there is that an empty list means _all domains_ which may not be obvious.

mkangia replied:
Thanks. 

If its okay, i am going to leave this as it is.
I don't see a valuable difference to what it is now and I prefer the quoted display for better grouping of text.

millerdev replied:
What about "misleading because it is not a valid value that can be used to reconstruct the object"?

mkangia replied:
I am not sure why that function will be used to reconstruct the object. I assume its only for display purpose?
And I believe it does not have any use really, i guess it was just added and can possibly be removed.
https://github.com/dimagi/commcare-hq/pull/17022/files

mkangia replied:
Ah, good to know. I was not aware.

As I shared earlier
> And I believe it does not have any use really, i guess it was just added and can possibly be removed.
https://github.com/dimagi/commcare-hq/pull/17022/files

I went ahead and removed the function as it does not seem to really getting used anywhere except for tests which I then modified.
https://github.com/dimagi/commcare-hq/pull/33592/commits/61e1cfe13d132030859b2d63f976c2fbb565b4bf

millerdev commented on corehq/apps/hqwebapp/models.py:
```diff
@@ -35,25 +35,29 @@ class MaintenanceAlert(models.Model):
     def save(self, *args, **kwargs):
-        MaintenanceAlert.get_active_alerts.clear(MaintenanceAlert)
-        super(MaintenanceAlert, self).save(*args, **kwargs)
+        CommCareHQAlert.get_active_alerts.clear(CommCareHQAlert)
+        super(CommCareHQAlert, self).save(*args, **kwargs)
```
nit
```suggestion
        cls = type(self)
        cls.get_active_alerts.clear(cls)
        super().save(*args, **kwargs)
```

mkangia replied:
Can you share why that is better?

millerdev replied:
It doesn't require updating the code if the class name changes, which makes the code more resilient over time.

In the case of the `super()` call, it's been unnecessary to pass arguments since Python 3.0.

millerdev commented on corehq/apps/hqwebapp/views.py:
```diff
@@ -423,6 +422,8 @@ def _login(req, domain_name, custom_login_page, extra_context=None):
         if couch_user:
             response.set_cookie(settings.LANGUAGE_COOKIE_NAME, couch_user.language)
+            # reset cookie to an empty list on login to show domain alerts again
+            response.set_cookie('viewed_domain_alerts', [])
```
Seems like this could be annoying to users who have a short login timeout.

millerdev commented on corehq/apps/hqwebapp/migrations/0011_auto_20231010_0950.py:
A more descriptive filename would be nice.

mkangia replied:
I have already deployed this on staging. I am happy to roll that back and redeploy after name change if you say its worth the effort.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/33592/commits/49c42c2180b1b88b6c0aff7f341cfe48c02f3bc7

millerdev commented on corehq/apps/domain/views/settings.py:
```diff
@@ -517,3 +520,87 @@ class ManageDomainMobileWorkersView(ManageMobileWorkersMixin, BaseAdminProjectSe
+    @cached_property
+    def form(self):
+        if self.request.method == 'POST':
+            return DomainAlertForm(self.request.POST)
```
For my own curiosity: the `text` field is the only thing that `DomainAlertForm` cares about, right? Does it matter if there are other fields present in `request.POST`?

mkangia replied:
Ya, that is the only field for now.
Unless someone adds something manually by modifying the HTML, `text` should be the only field.

Alert's do support a start/end time too which could be added in future if needed.

millerdev replied:
> Unless someone adds something manually by modifying the HTML, `text` should be the only field.

What about

https://github.com/dimagi/commcare-hq/blob/49c42c2180b1b88b6c0aff7f341cfe48c02f3bc7/corehq/apps/domain/views/settings.py#L553

and

https://github.com/dimagi/commcare-hq/blob/49c42c2180b1b88b6c0aff7f341cfe48c02f3bc7/corehq/apps/domain/views/settings.py#L565

millerdev replied:
Does that mean it is dangerous to reference `self.form` in this method as well as some methods that it calls (e.g., `_apply_command`) except in very particular circumstances? That was not obvious to me, and seems like an easy thing for future readers to overlook.

I suggest refactoring the code to make that not possible or at least much harder to do by mistake.

mkangia replied:
Not sure I understand. If someone calls `self.form` when it should not be, it would just give back a form with no data in it.

The entry function `post`, supports different operations (create or update/delete) and hence can get requests from different forms on the page. It decides which operations to do based on data passed in the request.

In order to make it more clear, is this better?
```
def create_alert_form()
    if self.request.method == 'POST':
        return DomainAlertForm(self.request.POST)
    return DomainAlertForm()

def post(self, request, *args, **kwargs):
    command = request.POST.get('command')
    if command:
        self._apply_command(request, command)
    else:
        if self.create_alert_form.is_valid():
            self._create_alert()
            messages.success(request, _("Alert saved!"))
        else:
            messages.error(request, _("There was an error saving your alert. Please try again!"))
            return self.get(request, *args, **kwargs)
    return HttpResponseRedirect(self.page_url)

def _apply_command(self, request, command):
    # ...
```
or you are proposing a structural change to the way the view and forms are interacting?
Like to just create the alert on the page via POST but handle activation/deactivation/deletion via a separate URL?
Or That a view should have only one form and that all operations should ideally be done by the same form?

millerdev replied:
I was thinking of something more like this
```py
def post(self, request, *args, **kwargs):
    command = request.POST.get('command')
    if command:
        self._apply_command(request, command)
    else:
        form = DomainAlertForm(request.POST)
        if form.is_valid():
            self._create_alert(form)
            ...
```
Where `form` is not a property of `ManageDomainAlertsView`. That way it cannot be referenced in an invalid context.

The reference in `page_context` still needs to be handled. It is unclear whether that method is ever called during a POST request that is not a `DomainAlertForm` submission, so unclear whether it was safe to reference `self.form` in there.

mkangia replied:
When there is an error in the form, the state of the form is to be retained so that the errors are displayed. A common way to do that is to have the form memoized, so when we return the `get` view, which will then call `page_context` and same form will be used.
So, until that is changed, `form` needs to be a property of `ManageDomainAlertsView`

millerdev replied:
To unblock this PR I'm going to drop this request for change.

My main complaint is that the context expected by the `form` property is dependent on things that are defined relatively distantly from the property. For example, it may not be obvious to a future developer who wants to change the `post` method that their change may cause `page_context` to access `self.form` when the submitted form does not match the fields of `DomainAlertForm`.

Said a slightly different way, it may not be obvious that the `post` method accepts multiple "different forms." This was not obvious to me.

mkangia replied:
Thanks for the relaxation @millerdev 

> it may not be obvious that the post method accepts multiple "different forms." This was not obvious to me.

This, however stuck with me and tried to make this better in the following commit from how I proposed an alternate earlier.
https://github.com/dimagi/commcare-hq/pull/33592/commits/a6808a8c9af4f974b8963880524d15737f8e0a06

The diff might not be easier to review so i left some comments. There is no functional feature change in that commit.
Please let me know if that makes the code clearer. Thanks for your patience with this.

millerdev replied:
That change did simplify `ManageDomainAlertsView.post`, but with the cost of adding two new URL endpoints. It's probably a nitpick  at this point, but I think it would have been cleaner to keep a single URL endpoint and use a different technique ensure that the logic in `ManageDomainAlertsView.form` could not be mistakenly called with the wrong form. However, I'm not sure how to better suggest a different approach apart from suggestions above. This solution is good enough.

millerdev reviewed (dismissed):
It does not look like tests were added for the new alerting functionality, especially around the error handling edge cases of the new view. Do you agree that they would be valuable?

Only reviewed python code, not JS or HTML changes.

mkangia commented on corehq/apps/hqwebapp/migrations/0011_auto_20231010_0950.py:
```diff
@@ -0,0 +1,31 @@
+        migrations.AddField(
+            model_name='maintenancealert',
+            name='created_by_domain',
+            field=models.CharField(db_index=True, max_length=255, null=True),
```
oh Cool. I was not aware of this. Just used what was used elsewhere for domains.

I will update and re-run the migration on staging.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/33592/commits/2c736cde3c48f2979c7d8202d8c04a395833eab2

mkangia commented:
@millerdev 

> It does not look like tests were added for the new alerting functionality, especially around the error handling edge cases of the new view. Do you agree that they would be valuable?

I could add tests to ensure that an alert updated/created/deleted on the view. However, I do believe those tests won't add much of a value to HQ test suite. The view is pretty simple, with one field in the form (marked as required) and then two buttons to activate/deactivate or delete the alert. So, it did feel that tests won't add any value here though I am happy to add them if it seems that they would. Or extract functions that modify alerts and test them without going through the view.

mkangia committed:
uses CharIdField for domain/userID columns

mkangia committed:
renames to a better name

mkangia committed:
fixes migration name

millerdev commented:
> I do believe those tests won't add much of a value to HQ test suite.

You modified 23 files and there are 26 commits in this PR. It took me quite some time to review it, and I didn't even review every file. From my perspective this is not a simple change, and tests would increase confidence that it works as intended and make it easier to maintain over time. Ideally there would be tests for all of the things mentioned in the number list in the safety story. However, happy path tests are often not enough to give much confidence that error conditions are handled correctly. I like to add tests for all edge cases I can think of that should have a well-defined result.

millerdev commented on corehq/apps/domain/views/settings.py:
```diff
@@ -517,3 +520,87 @@ class ManageDomainMobileWorkersView(ManageMobileWorkersMixin, BaseAdminProjectSe
+            return True
+
+    def _load_alert(self, alert_id):
+        try:
+            return CommCareHQAlert.objects.get(
+                created_by_domain=self.domain,
+                id=alert_id
+            )
+        except CommCareHQAlert.DoesNotExist:
+            return None
```
```suggestion
        return CommCareHQAlert.objects.get(
            created_by_domain=self.domain,
            id=alert_id
        )
```

mkangia replied:
Definitely letting an exception get raised isn't an ideal experience.

Added a notification to user in case alert isn't found
https://github.com/dimagi/commcare-hq/pull/33592/commits/73f9a875ba16b83557c65ba5e406843a626dd86d

mkangia committed:
refactor

1. avoid using class name
2. avoid passing unnecessary arguments to super call

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

mkangia committed:
lets user know if alert isn't found

mkangia committed:
lets user know if unexpected command is received

mkangia commented:
@millerdev 

> From my perspective this is not a simple change, and tests would increase confidence that it works as intended and make it easier to maintain over time. Ideally there would be tests for all of the things mentioned in the number list in the safety story. However, happy path tests are often not enough to give much confidence that error conditions are handled correctly. I like to add tests for all edge cases I can think of that should have a well-defined result.

Thanks for reiterating on this. You are right. I underestimated the importance of test here. Seems like something I need to update generally for my understanding as well. I will add them now.

> You modified 23 files and there are 26 commits in this PR. It took me quite some time to review it, and I didn't even review every file. 

I didn't think about the number of commits and files changed. I don't believe those numbers reflect correctly about the change being made.

Generally, I do like to break down changes into smaller commits so the number of commits is mostly higher for me though this PR had iterations on the approach to rollout regarding the table name, which added some overhead commits. So initially, review by commit made sense but may be it does not anymore.

Thanks for the review Daniel. I appreciate your effort for the review on this PR.

mkangia committed:
adds tests for view

mkangia committed:
asserts for critical parameter

mkangia committed:
adds tests for restrictions to maintenance alerts view

mkangia committed:
adds tearDown for class

millerdev reviewed (commented):
Thanks for adding tests.

mkangia committed:
removes redundant method

mkangia committed:
skips cleanup for sql alerts

mkangia committed:
improves clean up by using addsClassCleanup

mkangia commented on corehq/apps/domain/tests/test_views.py:
```diff
@@ -111,6 +115,169 @@ def test_autocomplete_disabled(self):
+    @flag_enabled('CUSTOM_DOMAIN_BANNER_ALERTS')
+    def test_apply_command_with_valid_command(self):
+        alert = self._create_alert_for_domain(self.domain, "New Alert!", self.username)
+        self.addCleanup(alert.delete)
```
Cool!

https://github.com/dimagi/commcare-hq/pull/33592/commits/33d3e57763e11ec04685e8926f87bddf3972a19a


My understanding was that,
when we run tests locally where the tests don't stop on failure, this cleanup would be relevant.
I believe that if this test would fail the alert created will be retained and could conflict with another test.

millerdev replied:
> I believe that if this test would fail the alert created will be retained and could conflict with another test.

Django's [`TestCase`](https://docs.djangoproject.com/en/4.2/topics/testing/tools/#testcase) will always roll back SQL database state, regardless of whether it passes or fails, running locally or on Github Actions (assuming the database backend is one like PostgreSQL that supports transactions).

mkangia replied:
Ah, thank you.
I had it wrong.

mkangia commented on corehq/apps/domain/tests/test_views.py:
```diff
@@ -111,6 +115,169 @@ def test_autocomplete_disabled(self):
+        cls.domain_alert.delete()
+        cls.other_domain_alert.delete()
+        cls.user.delete(cls.domain_name, deleted_by=None)
+        cls.domain.delete()
```
oh cool, I thought that is how we would do clean ups generally on HQ for objects created in `setUpClass`.

https://github.com/dimagi/commcare-hq/pull/33592/commits/13771605bd182796cdbb2bc1c023946b884c0556

mkangia committed:
renames model

mkangia committed:
makes separate views for update and delete

mkangia commented on corehq/apps/hqwebapp/migrations/0012_rename_commcarehqalert_alert.py:
```diff
@@ -0,0 +1,17 @@
+        migrations.RenameModel(
+            old_name='CommCareHQAlert',
+            new_name='Alert',
+        ),
```
```
$ ./manage.py sqlmigrate hqwebapp 0012
BEGIN;
--
-- Rename model CommCareHQAlert to Alert
--
COMMIT;
```

mkangia commented on corehq/apps/domain/tests/test_views.py:
```diff
@@ -170,6 +176,64 @@ def test_only_domain_alerts_listed(self):
         self.assertEqual(response.status_code, 200)
 
+    @flag_enabled('CUSTOM_DOMAIN_BANNER_ALERTS')
+    def test_creating_new_alert(self):
```
tests in this class are same as how they were before.

mkangia commented on corehq/apps/domain/tests/test_views.py:
```diff
@@ -170,6 +176,64 @@ def test_only_domain_alerts_listed(self):
+        self.assertEqual(response.status_code, 404)
+
+    @flag_enabled('CUSTOM_DOMAIN_BANNER_ALERTS')
+    def test_post_access_only(self):
```
this is a new test added under this view since it enforces POST

mkangia commented on corehq/apps/domain/tests/test_views.py:
```diff
@@ -241,44 +305,56 @@ def test_apply_command_with_other_doamin_alert(self):
     @flag_enabled('CUSTOM_DOMAIN_BANNER_ALERTS')
-    def test_creating_new_alert(self):
-        self.assertEqual(Alert.objects.count(), 2)
+    def test_post_access_only(self):
```
this is a new test added under this view since it enforces POST

mkangia commented on corehq/apps/domain/tests/test_views.py:
```diff
@@ -241,44 +305,56 @@ def test_apply_command_with_other_doamin_alert(self):
+            )
+
+    @flag_enabled('CUSTOM_DOMAIN_BANNER_ALERTS')
+    def test_with_missing_alert(self):
```
simply added tests for missing alert and a successful delete after this but due to the diff, it looks like I updated an existing test.

mkangia commented on corehq/apps/domain/views/settings.py:
```diff
@@ -550,54 +551,75 @@ def form(self):
+    return HttpResponseRedirect(reverse(ManageDomainAlertsView.urlname, kwargs={'domain': domain}))
+
+
+def _load_alert(alert_id, domain):
```
Just extracted out of class, definition same as before.

mkangia commented on corehq/apps/domain/views/settings.py:
```diff
@@ -550,54 +551,75 @@ def form(self):
+        return None
+
+
+def _apply_command(request, alert, command):
```
Extracted out of class and removed support for delete command since now its a separate view of its own.

mkangia replied:
Additionally, the check for alert_id was moved up in the views itself.

mkangia commented on corehq/apps/domain/views/settings.py:
```diff
@@ -550,54 +551,75 @@ def form(self):
+        messages.error(request, _("Unexpected update received. Alert not updated!"))
+
+
+def _update_alert(alert, command):
```
Just extracted out of class, definition same as before.

mkangia committed:
update model name for domain delete

mkangia committed:
fixes naming

mkangia committed:
displays user who created the alert for maintenance alerts

mkangia committed:
fix test

mkangia committed:
sets up method to check access to domain

- add ensure_requet_project just like for couch_user
- extract active_user_logged_in

mkangia committed:
loads domain alerts only when accessible

mkangia commented:
Hey @millerdev 

I just a made a small separate PR for the QA fixes for this feature change for easier review. It's quite a separate change so should be possible to review it without needing to refer changes in this PR.
https://github.com/dimagi/commcare-hq/pull/33644

fyi @kaapstorm (I know you are away for a bit, but just FYI)

millerdev commented on corehq/apps/domain/views/settings.py:
```diff
@@ -550,54 +551,75 @@ def form(self):
+    if not alert:
+        messages.error(request, _("Alert not found!"))
+    else:
+        command = request.POST.get('command')
+        if command:
+            _apply_command(request, alert, command)
```
The result would appear to be more consistent without the conditional
```suggestion
        command = request.POST.get('command')
        _apply_command(request, alert, command)
```

mkangia committed:
updates comments to docstring

mkangia committed:
avoids code duplication

mkangia committed:
lint

mkangia committed:
simplifies & includes error message for missing command as well

mkangia committed:
merges migrations

mkangia committed:
Merge branch 'master' into mk/3003-add-domain-banners

mkangia committed:
Merge pull request #33644 from dimagi/mk/3003-add-domain-banners-qafix

Fixes for QA for custom domain banners

mkangia committed:
updates migrations.lock for removed migration

mkangia committed:
corrects function path

mkangia committed:
Merge branch 'master' into mk/3003-add-domain-banners

mkangia commented on migrations.lock:
```diff
@@ -594,7 +594,6 @@ hqwebapp
  0009_truncate_authtoken_table
  0010_maintenancealert_scheduling
  0011_add_new_columns_and_rename_model
- 0012_rename_commcarehqalert_alert
```
Missed updating this in commit earlier, https://github.com/dimagi/commcare-hq/pull/33592/commits/027ead6bfd418949894569bd73b18cd6867b8106

mkangia committed:
corrects function path

mkangia committed:
adds missing super calls

millerdev commented on corehq/apps/hqwebapp/tests/test_views.py:
```diff
@@ -144,6 +144,7 @@ def setUpClass(cls):
     def tearDownClass(cls):
         cls.user.delete(cls.domain, deleted_by=None)
         cls.project.delete()
```
It would be safer to use `cls.addClassCleanup` to delete these.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/33592/commits/119d15c5c54623e2ceb32a2091801a1751272811

mkangia committed:
improves test setup
